### DISCUSSION TITLE
spatial gradient of material

### DISCUSSION THREAD

**Discussion by mortezaaesmaeilpour** [idaholab/moose#20628] [Q&A General] [answered]
Dear MOOSE team

as I know it is easy to use spatial derivatives of main variables (EX: _grad_p(coupledGradient("pressure"))). Is it also possible to easily calculate and use spatial derivatives of material components (like _grad_viscosity)? I know it is possible to calculate the derivative of material over the main variable (https://mooseframework.inl.gov/source/materials/DerivativeMaterialInterface.html). But, here my question is about SPATIAL derivatives, like a derivative of viscosity over X, Y, Z.

Kind regards
Morteza

**Comment by GiudGiud** [comment, marked as answer]
Hello

So there are two options:
1) Define it yourself in a new material property. 
`grad_mu = dmu/dT grad_T + dmu/dP grad_P`
you can create a new material that does exactly that.

2) Hack the derivative material interface system, using 'x' as a variable or a material property.
I ll post how to do this later, got to check something first

Guillaume

**Reply by mortezaaesmaeilpour** [reply]
Yes, I was looking for the second option.

**Reply by simopier** [reply]
Hi @GiudGiud, 

I am currently trying to do something similar in a kernel that would uses the gradient of the hydrostatic stress. The hydrostatic stress is provided as a material property by the tensor mechanics module. Since it can be affected by many variables (displacements, temperature, pressure, etc.), defining its gradient in a new material property can be challenging. 

In [this issue](https://github.com/idaholab/moose/discussions/18415), you hinted at some ongoing development that would allow this. Do you have any updates on that?

Thank you!

Best,
PC

**Reply by GiudGiud** [reply]
@simopier I was mentioning the functor material property system. It's actually merged in MOOSE. It evaluates material properties (and their gradient, time derivatives etc) on demand, at arbitrary locations.
It's quite easy to use but it's not pluggable in the interface for regular material properties. I can point you to some examples if you want to have a look. We currently only use it in the Navier Stokes module.

@mortezaaesmaeilpour no luck on DerivativeParsedMaterial, it seems to only do symbolic derivations, gives me 0 for the derivatives.

If you dont go with 1., you have two other options to compute gradients of material properties:
- the functor material system, see this documentation.
https://mooseframework.inl.gov/syntax/Materials/#e36f00e4-37cd-45d4-a2f6-61c87f218c47
It'll require some changes in all your material properties.

- using automatic differentiation and retrieving the derivative (wrt to a main non linear variable) from the AD number, see this documentation https://mooseframework.inl.gov/moose/modules/phase_field/FunctionMaterials/AutomaticDifferentiation.html

**Comment by ngrilli** [comment]
Dear @mortezaaesmaeilpour 

In the past, I did the following:
1) define a first monomial AuxVariable, representing the material property that you want to derive 
2) use the AuxKernel MaterialStdVectorAux to assign to the first monomial variable the material property
3) use VariableGradientComponent to calculate its derivative and assign it to another constant monomial AuxVariable
4) getMaterialProperty in kernels or materials to import the value of the derivative.

You can see an example here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/SandfeldZaiser2015/SandfeldLoop.i
the auxvariable of interest is called "dislov" and "ddislovdx" is the derivative.

I am looking forward to hear from @GiudGiud  about better strategies to do this.

Best Regards,
Nicolò

**Reply by mortezaaesmaeilpour** [reply]
Hi @ngrilli 

Do you think It is a good idea to convert a material property to an AuxVariavle and use the spatial derivative of this AuxVariable for the calculations of redial and Jacobians in kernels? I mean the material properties and main variables are supposed to be calculated simultaneously. However, AuxVariables are not included in the main solution matrix.

Kind regards
Morteza

**Reply by ngrilli** [reply]
Dear @mortezaaesmaeilpour 

It depends on your specific problem.
A material property is meant to represent a quantity that can be derived from the problem variables.
Sure material properties and auxvariables are not solved for, in the FEM sense, as you said.

You can always substitute material properties with variables and kernels for their evolution equation,
but this is not always necessary, it depends on the problem.
If it is a "good" or "bad" idea depends on the numerical problem you are solving for.

I normally work with plasticity models, in which plastic deformation is set as a material property
and updated explicitly. Of course, it would be always possible to add it as a variable instead,
but that would make the simulation much slower without giving any improvement in the precision.

It's up to your specific problem and how people have addressed it in the literature.
A material property or auxvariable is something you calculate while variables are kept fixed in a time step.
On the other hand, variables are updated while material properties are kept fixed.
You can think about it as a staggered scheme.

Best Regards,
Nicolò
